hg_agent_forwarder/forwarder.py

```diff
@@ -43,27 +42,67 @@ def __init__(self, config, *args, **kwargs):
         self.backoff = False
         self.request_session = requests.Session()
         self.request_session.auth = HTTPBasicAuth(self.api_key, '')
+        self.batch = ""
+        self.batch_size = 0
+        self.batch_time = time.time()
+        self.batch_timeout = config.get('batch_timeout', 0.5)
+        self.max_batch_size = config.get('max_batch_size', 250)
 
     def run(self):
         while not self.shutdown_e.is_set():
             try:
                 for line in self.spool_reader.read():
                     datapoint = Datapoint(line, self.api_key)
                     if datapoint.validate():
-                        self.forward(datapoint)
+                        self.extend_batch(datapoint)
                     else:
                         logging.error("Invalid line in spool.")
                         # invalid somehow, pass
                         continue
+                    if self.should_send_batch():
+                        self.forward()
                 if self.shutdown_e.is_set():
                     break
             except Exception as e:
                 continue
 
-    def forward(self, data):
+    def extend_batch(self, data):
+        '''
+        Add a metric to the current metric batch.
+        '''
+        if not self.batch_time:
+            self.batch_time = time.time()
+
+        try:
+            metric = data.metric
+            value = data.value
+            ts = data.timestamp
+        except AttributeError:
+            # somehow, this dp is invalid, pass it by.
+            return True
+
+        metric_str = "%s %s %s" % (metric, value, ts)
+        self.batch = "%s\n%s" % (self.batch, metric_str)
+        self.batch_size += 1
```

Comment:
We should have this return something or change the exception handling return to not return True.